fendermint/vm/topdown/src/sync/syncer.rs

```diff
@@ -0,0 +1,268 @@
+            Ok(res) => res,
+            Err(e) => {
+                let err = e.to_string();
+                if is_null_round_str(&err) {
+                    tracing::warn!(height, "null round at height");
+
+                    self.sync_pointers.advance_head();
+
+                    return Ok(None);
+                }
+                return Err(Error::CannotQueryParent(err, height));
+            }
+        };
+
+        if block_hash_res.parent_block_hash != parent_block_hash {
+            tracing::warn!(
+                height,
+                parent_hash = hex::encode(&block_hash_res.parent_block_hash),
+                previous_hash = hex::encode(&parent_block_hash),
+                "parent block hash diff than previous hash",
+            );
+            return Err(Error::ParentChainReorgDetected);
+        }
+
+        if let Some(to_confirm) = self.sync_pointers.to_confirm() {
+            tracing::warn!(
+                height,
+                confirm = to_confirm,
+                "non-null round at height, confirmed previous height"
+            );
+            let data = self
+                .fetch_data(to_confirm, block_hash_res.block_hash)
+                .await?;
+            self.sync_pointers.advance_confirm(height);
+            return Ok(Some((to_confirm, data)));
+        }
+
+        tracing::warn!(height, "non-null round at height, waiting for confirmation");
+        self.sync_pointers.advance_head();
+
+        Ok(None)
+    }
+
+    async fn fetch_data(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+    ) -> Result<ParentViewPayload, Error> {
+        let changes_res = self
+            .parent_proxy
+            .get_validator_changes(height)
+            .await
+            .map_err(|e| Error::CannotQueryParent(e.to_string(), height))?;
+        if changes_res.block_hash != block_hash {
+            tracing::warn!(
+                height,
+                change_set_hash = hex::encode(&changes_res.block_hash),
+                block_hash = hex::encode(&block_hash),
+                "change set block hash does not equal block hash",
+            );
+            return Err(Error::ParentChainReorgDetected);
+        }
+
+        let top_down_msgs_res = self
+            .parent_proxy
+            .get_top_down_msgs_with_hash(height, &block_hash)
+            .await
+            .map_err(|e| Error::CannotQueryParent(e.to_string(), height))?;
+
+        Ok((block_hash, changes_res.value, top_down_msgs_res))
+    }
+
+    /// We only want the non-null parent block's hash
+    async fn non_null_parent_hash(&self) -> BlockHash {
+        let parent_height = match (self.sync_pointers.to_confirm(), self.sync_pointers.tail()) {
+            (Some(height), _) => {
+                tracing::debug!(height, "found height to confirm");
+                height
+            }
+            (None, height) => {
```

Comment:
This will only happen in the beginning right? According to https://github.com/consensus-shipyard/fendermint/pull/442/files#diff-7f8fecef12fedb16dfaea5611b9eccaf7be6b3ffa36059933234d77b74310534R172-R183, after the first block this should always be `Some`. 

Also, `advance_confirmed` is only called there, so I don't know to what extent `to_confirm` will ever be set to some other value different `None`. I am afraid there may be a bug there.